Inductor reduce_scatter_tensor

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #95764

This adds reduce_scatter to the functional collective and adds the
inductor lowering support

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

wanchaol committed:
Inductor reduce_scatter_tensor

This adds reduce_scatter to the functional collective and adds the
inductor lowering support

[ghstack-poisoned]

awgu commented on torch/_inductor/ir.py:
```diff
@@ -4363,3 +4363,59 @@ def codegen(self, wrapper):
+
+        # Collective returns a 'work' object.  But Inductor's scheduler doesn't need to know
+        # about that, and we just pretend for scheduling purposes that the work obj is a 1-elem tensor.
+        # Nobody should consume the output of AllReduce except 'Wait', which we control here.
```
nit: For this PR and also for the `all_gather_into_tensor` inductor PR (see [here](https://github.com/pytorch/pytorch/pull/95530/files#diff-cf6ca00beddc32a2a6a2933fb9913b6a2b925ffc3b745488967210e4343134acR4334-R4336)), could we fix the comments to not mention "AllReduce"? This seems like possibly from a copy and paste.

wconstab replied:
you're not supposed to read the comments :P  

(yes, lets  fix!)

wconstab commented on torch/distributed/_functional_collectives.py:
```diff
@@ -132,13 +139,15 @@ def unwrap(e: Any):
         out = func(*unwrapped_args, **unwrapped_kwargs)
         return out
 
+
```
wanchao are you on a different version of linter than me? why are there so many formatting changes?

prefer a separate pr to update formatting if necessary

wanchaol replied:
seems weird as I didn't turn on the auto format on save feature... let me revert those changes

wanchaol replied:
probably because I enabled ufmt in my editor and that seems applied autoformatting lol, reverted the format changes back

wanchaol committed:
Update on "Inductor reduce_scatter_tensor"


This adds reduce_scatter to the functional collective and adds the
inductor lowering support

cc soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

wanchaol committed:
Update on "Inductor reduce_scatter_tensor"


This adds reduce_scatter to the functional collective and adds the
inductor lowering support

cc soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

wanchaol committed:
Update on "Inductor reduce_scatter_tensor"


This adds reduce_scatter to the functional collective and adds the
inductor lowering support

cc soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

wconstab commented on aten/src/ATen/native/Collectives.cpp:
```diff
@@ -24,6 +24,10 @@ at::Tensor all_gather_into_tensor(at::Tensor const& shard, const c10::string_vie
     TORCH_INTERNAL_ASSERT(false);
 }
 
+at::Tensor reduce_scatter_tensor(at::Tensor const& shard, const c10::string_view reduceOp, int64_t scatter_dim, const c10::string_view tag, c10::ArrayRef<int64_t> ranks, int64_t group_size) {
```
is the naming correct?  shard is the input to reduce_scatter? or shard is the output and ... idk, source, tensor, etc is the input?

wanchaol replied:
Sorry was a copy-pasta issue haha, I think maybe we can just use "input" to align with the eager API naming? Updated

wanchaol committed:
Update on "Inductor reduce_scatter_tensor"


This adds reduce_scatter to the functional collective and adds the
inductor lowering support

cc soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

wanchaol committed:
Update on "Inductor reduce_scatter_tensor"


This adds reduce_scatter to the functional collective and adds the
inductor lowering support

cc soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

wanchaol committed:
Update on "Inductor reduce_scatter_tensor"


This adds reduce_scatter to the functional collective and adds the
inductor lowering support

cc soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

kumpera commented on torch/distributed/_functional_collectives.py:
```diff
@@ -165,6 +168,30 @@ def _all_gather_into_tensor(shard, tag, ranks, group_size):
+    ranks: List[int],
+    group_size: int,
+):
+    # TODO add dim support?
```
I've being thinking about dim support and for it to work we'd need to change the strides of the input and rely on the compiler to propagate that back all the way to the the DTensor local_shard.

It's a nice project on itself, but one we can do later.

wanchaol replied:
yeah changing stride requires a copy of the tensor as nccl needs contiguous buffer, we need to think about how to deal with it in the reduce_scatter tensor API, otherwise DTensor can't use (it currently uses the list option as it allows catting back on a different dimension)

kumpera reviewed (approved):
LGTM

wanchaol commented:
@pytorchbot merge
